Username to be displayed instead of user id with clickable link

InJoDave opened the pull request:
Replaced userid with username as column value to be displayed in User Events tab

Fixes : https://github.com/keycloak/keycloak/issues/39357

Signed-off-by: Indu John [john.indu@gmail.com](mailto:john.indu@gmail.com)

InJoDave committed:
Username to be displayed instead of user id with clickable link

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave commented:
@mpawlus for your review, requested changes done in this PR

mpawlus commented on server-spi-private/src/main/java/org/keycloak/events/Event.java:
```diff
@@ -102,6 +104,12 @@ public void setUserId(String userId) {
+    public String getUserName() { return userName; }
+
+    public void setUserName(String userName) {
+        this.userName = maxLength(userName,255);
```
Please fix the formatting:
`this.userName = maxLength(userName, 255);`

InJoDave replied:
ok

mpawlus commented on server-spi-private/src/main/java/org/keycloak/events/EventBuilder.java:
```diff
@@ -127,6 +127,11 @@ public EventBuilder client(String clientId) {
 
     public EventBuilder user(UserModel user) {
         event.setUserId(user == null ? null : user.getId());
+        if(user!=null && user.getUsername()!=null) {
```
Same here:
`if (user!=null && user.getUsername() != null) {`

mpawlus commented on model/jpa/src/main/java/org/keycloak/events/jpa/EventEntity.java:
```diff
@@ -48,6 +48,9 @@ public class EventEntity {
     @Column(name="USER_ID")
     private String userId;
 
+    @Column(name="USER_NAME",nullable = true)
```
Does it make sense to have no username?
Even if, then the formatting needs to be fixed:
`@Column(name="USER_NAME", nullable = true)`

InJoDave replied:
I added this so that in case there is an event where username is null, it doesn't block the application. Will fix will formatting issues

mpawlus commented on model/jpa/src/main/java/org/keycloak/events/jpa/EventEntity.java:
```diff
@@ -112,6 +115,12 @@ public void setUserId(String userId) {
+        return userName;
+    }
+
+    public void setUserName(String userName) { this.userName = userName; }
```
To stay consistent, I would put the code in a separate line, just like for other setters.

InJoDave commented:
@mpawlus let me try this approach and push the changes.

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

InJoDave committed:
fix for test failing for username column added

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

InJoDave committed:
fix for test failing for username column added

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave commented:
@mpawlus Could you please check the changes now? There are some failing for which I need some help

InJoDave committed:
fix for test failing for username column 2

Signed-off-by: Indu John <john.indu@gmail.com>

ahus1 committed:
More eagerly clear references to client sessions from the user session (#39652)

Closes #39651

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Send events in batch with the new method (#39654)

Closes #39653

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

mpawlus commented:
@InJoDave I created a PR with some fixes, please have a look: https://github.com/InJoDave/keycloak/pull/1

InJoDave commented:
@mpawlus I think I just did same correction...but let me recheck if I missed it

InJoDave commented:
@mpawlus can you recheck as we both did same changes in the same Test classes

mpawlus commented:
> @mpawlus can you recheck as we both did same changes in the same Test classes

@InJoDave It looks like 2 lines were missed, I think you can merge my PR.

mpawlus commented:
@InJoDave My bad, I did not sign-off my commit. Once all the checks pass, can you please fix it as per https://github.com/keycloak/keycloak/pull/39640/checks?check_run_id=42125693313? Thanks!

InJoDave commented:
@mpawlus I think you can edit commit message in IntelliJ , this is what I used to do

edewit committed:
use the selected options in the render (#39535)

fixes: #39496

Signed-off-by: Erik Jan de Wit <erikjan.dewit@gmail.com>

ahus1 committed:
The transport stack defined in the cache XML should take precedence over internal jdbc-ping default

Closes #39614

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Lazily process sessions from ISPN to avoid fetching client sessions (#39639)

Closes #39638

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Reorder operations to avoid the slow operation to get all client sessions (#39666)

Closes #39665

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

mpawlus commented:
> @mpawlus I think you can edit commit message in IntelliJ , this is what I used to do

But I won't be able to force push to this repo (unless you make me a contributor here, which think is an overkill).

ronisaha committed:
Remove redundant username error display logic

Closes #39663

The error message rendering for the username field was removed as it is already handled in @field.input macro

Signed-off-by: Roni Saha <roni.cse@gmail.com>

InJoDave commented:
@mpawlus share your sign of message, will try to edit from my side

mpawlus commented:
> Signed-off-by: Indu John

@InJoDave You can use this one:
Signed-off-by: Maciej Pawlus m.t.pawlus@gmail.com

mpawlus commented:
> @mpawlus share your sign of message, will try to edit from my side
@InJoDave You can use this one:
Signed-off-by: Maciej Pawlus [m.t.pawlus@gmail.com](mailto:m.t.pawlus@gmail.com)

ahus1 reviewed (dismissed):
I see this PR adds the username to the table. Usually we only add column to the database when there is a need to search for that data. All other information is then stored in the details_json. 

Is there a particular need to add this column? As this table is usually very large, the migration of the table will take quite some time for some users of Keycloak, so I wonder if it can be avoided.

InJoDave commented:
> I see this PR adds the username to the table. Usually we only add column to the database when there is a need to search for that data. All other information is then stored in the details_json.
> 
> Is there a particular need to add this column? As this table is usually very large, the migration of the table will take quite some time for some users of Keycloak, so I wonder if it can be avoided.

As requested username was also suggested to be added into User Events. So adding it in the DB as new columns seemed the logical solution instead of overwriting any existing column. In the dashboard but we are displaying this value instead of userid which is already reviewed by @edewit 
Let me know what change is required here

mpawlus committed:
fix: adjust tests

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus committed:
fix: adjust tests

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus commented:
@InJoDave I have pushed rebased commits, now the issue with sign-off is gone. Will you take care of changes requested by @ahus1?

InJoDave commented:
@mpawlus yes I can do it but I didnt understand what needs to be done? I have given reply as to why I added username because of course you would need capability to search by username also in events as per my understanding. If you want to work on that I am ok with it also as already you are added as contributor to this branch

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus committed:
fix: adjust integration tests

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus commented:
@ahus1 I agree with @InJoDave that the functionality of searching users by username is a must. Taking this into consideration, what's your suggestion?

ahus1 commented:
In Keycloak, a user has both a username and and a user id. The username can change over time, while the user id remains constant. 

Reading the original issue, it is about showing the username and making it clickable. It is not about searching in the UI for the username. Testing this PR, searching is still be user ID. 

When searching for a user, I would expect to find all events of that user, even before it was renamed. So searching the database by username is IMHO not a valid option.

When search with a username, one would first look up the user by its user name to retrieve its user id, and then use the user id to search the database to find all events before and after the user name rename. 

At the same time a user might have been deleted in the meantime, so the only way to look up the events connected to that now deleted user is the user id. So searching by user id should always be supported, with searching by username is optional, and should then do it only after looking up the user id. 

As the same time it would be nice to have the username in the event already to have a better view for humans. Still to note that a username shown in the events might already have been deleted and possibly replaced by a different user with the same name. And when searching the events for a given user id, the username shown might change in the table view if the user has been renamed (so you would have old entries with the old username, and new entries with a new username). 

So I'd also argue to keep the username as a part of the details of the event. I agree with shouldn't set it manually in all the places, but rather automatically every time we set the user.

At the same time, not storing the username as before in the event would be a breaking change, which we would need to wait for a major release of Keycloak. And even then, people would complain a lot, and IMHO it is an avoidable change.
Still I like the idea of adding the username to the event every time we set the UserModel in the event builder. This would make it a lot more consistent.

mpawlus commented:
@ahus1 Ok, so if I get it right the idea is to restore the username as part of the event details (i.e. reverting it from the event directly) and making sure it is always present when building the event? @InJoDave  If so, then for the sake of clarity, I think we should close this pull request and open a new one.

ahus1 commented:
@mpawlus - yes, that what I suggest. Let's hear what the others say about it. I'd say it can still be this PR, and once can then squash any commits later. We would still have the discussion in one place if we keep this PR.

InJoDave commented:
@mpawlus I agree with @ahus1 about keeping same PR as already this much effort I have put into it and a good insightful conversation of improving the same is already going on here. Also I can take up this @mpawlus

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

Signed-off-by: InJoDave <75904306+InJoDave@users.noreply.github.com>

InJoDave commented:
@ahus1 This is what I suggested where the username is actually being ready from details only and shown alongwith userid. So in case as you said username was changed the next event with new username will have username but also userid to identify that both are same users. Hope this is ok with @edewit as well? 

![image](https://github.com/user-attachments/assets/ee57ae76-019c-4cd2-88cd-77e7c75f277d)

InJoDave commented:
@ahus1 but I see that username is not editable. Could you please let me know how it can be done so that I can test this as well
![image](https://github.com/user-attachments/assets/a7ac1d55-894f-4bc6-9b65-2b5ba372aeb1)

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

mpawlus commented on core/src/main/java/org/keycloak/representations/idm/EventRepresentation.java:
```diff
@@ -30,6 +30,7 @@ public class EventRepresentation {
     private String realmId;
     private String clientId;
     private String userId;
+    private String userName;
```
@InJoDave  Shouldn't this be reverted as well?

InJoDave replied:
yes, sorry missed it. Will commit with test changes now

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

mpawlus commented on model/jpa/src/main/resources/META-INF/jpa-changelog-26.3.0.xml:
```diff
@@ -15,12 +15,12 @@
   -->
-<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
-
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
@InJoDave I think this file should be reverted as well.

InJoDave replied:
@mpawlus I did revert my changes, the other table mentioned is added from master, not my change

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

mpawlus commented on services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:
```diff
@@ -1201,7 +1200,7 @@ public Response finishAuthentication(LoginProtocol protocol) {
     public void validateUser(UserModel authenticatedUser) {
         if (authenticatedUser == null) return;
         if (!authenticatedUser.isEnabled()) {
-            event.user(authenticatedUser).detail(Details.USERNAME, authenticatedUser.getUsername());
```
@InJoDave Shouldn't this be revert this as well?

InJoDave replied:
yes

mpawlus commented on js/apps/admin-ui/src/events/UserEvents.tsx:
```diff
@@ -92,23 +95,28 @@ const DetailCell = (event: EventRepresentation) => (
+      >
+        {userName}
+      </Link>
+      <span style={{ fontSize: '0.9rem', color: '#888', marginLeft: '5px' }}>
```
Let's get rid of this formatting errors too.

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave commented:
@ahus1 Could you please review the changes? I have attached screenshot also

mpawlus commented on js/apps/admin-ui/src/events/UserEvents.tsx:
```diff
@@ -92,23 +95,28 @@ const DetailCell = (event: EventRepresentation) => (
+      >
+        {userName}
+      </Link>
+      <span style={{ fontSize: '0.9rem', color: '#888', marginLeft: '5px' }}>
```
@InJoDave  Can you please fix both formatting issues, as they fail one of the build checks?

InJoDave replied:
done

InJoDave committed:
Merge branch 'keycloak:main' into add-username-to-events-updated

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

InJoDave committed:
Suggested changes to remove username from Events table and use details as before and propagate its value automatically in EventBuilder same like userid

Signed-off-by: Indu John <john.indu@gmail.com>

ahus1 commented:
@InJoDave - unfortunately I will not have the time to review this PR in the next two weeks. I will come back to it later. I dismissed my previous comment as the changes have been addressed to unblock this in case someone else wants to review it.

InJoDave commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedActionTotpSetupTest.java:
```diff
@@ -102,7 +102,7 @@ public void setupTotpRegister() {
         loginPage.clickRegister();
         registerPage.register("firstName", "lastName", "email@mail.com", "setupTotp", "password", "password");
 
-        String userId = events.expectRegister("setupTotp", "email@mail.com").assertEvent().getUserId();
```
Sure. Because I am using username from the "UserModel" which is always stored in lowercase which is set by the UserAdapter as given below 
![image](https://github.com/user-attachments/assets/269573bb-e035-4a61-9fdd-e3d2bc519299)

mpawlus committed:
fix: adjust styles for user id section in event details

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus commented:
@InJoDave I think I might have a fix for one of the issues, but I can't push anymore, I keep getting 403 response - can you please double-check if I am still a contributor?

InJoDave commented:
> @InJoDave I think I might have a fix for one of the issues, but I can't push anymore, I keep getting 403 response - can you please double-check if I am still a contributor?

I have added you @mpawlus as well as @ahus1 as contributor. You can merge the test changes @mpawlus. Sorry for late reply

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

InJoDave committed:
Merge branch 'main' into add-username-to-events-updated

InJoDave commented:
@mpawlus / @ahus1 is this still needing some change?

mpawlus committed:
fix: adjust syntax for setting class name

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
refactor: restore changelog formatting

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

ahus1 committed:
A username already specified in the event should take precedence

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Username like this should be preserved

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
adding a blank between the values, and removing "ID:"

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented on services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:
```diff
@@ -1180,7 +1180,6 @@ public static ClientSessionContext attachSession(AuthenticationSessionModel auth
         }
 
         event.user(userSession.getUser())
-                .detail(Details.USERNAME, username)
```
This is an example where the username should be stored as entered by the user, as we don't want to change the logic of existing events. It might sometimes be important to know what the user actually entered in the form. 

So this line and similar ones should be restored.

ahus1 commented on services/src/main/java/org/keycloak/authentication/requiredactions/DeleteAccount.java:
```diff
@@ -107,7 +106,6 @@ public void processAction(RequiredActionContext context) {
         eventBuilder.event(EventType.DELETE_ACCOUNT)
             .client(keycloakContext.getClient())
             .user(user)
-            .detail(Details.USERNAME, user.getUsername())
```
Removing this line is fine, as here the username is picked from the user, identical to what we do when constructing the event when passing in the user.

ahus1 commented on js/apps/admin-ui/src/events/UserEvents.tsx:
```diff
@@ -92,23 +95,26 @@ const DetailCell = (event: EventRepresentation) => (
-      )}
-      {!event.userId && t("noUserDetails")}
-    </>
+  return userId && userName ? (
```
This change now requires both userId and userName to be set. But there are a lot of old events in the database that have a userId, but no username. And those are connected to a user. Those should be displayed with the userID and should link to the user (as before).

ahus1 reviewed (changes requested):
Thank you for the changes, and sorry for the wait. 

I had a look and when I scrolled to the end of the PR, I saw that several test cases are changed as now a different username is recorded. 

IMHO this would be a breaking change to those who are parsing the events, and that want to see how a user logged in: either with a username, or an email address, etc. 

See below for one example where adding the username can be safely removed, and another example where  to keep the username.

As a test, all changes to the test cases that relate to a change in the username should then be reverted as well.

mpawlus committed:
fix: restore setting username as part of user details

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
fix: handle missing username in user events section

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge remote-tracking branch 'upstream/main' into add-username-to-events-updated

mpawlus committed:
chore: remove styling for obsolete class

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
refactor: fix Maven spotless issue

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
refactor: fix indentation issues from prettier

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
test: revert changes in tests

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
test: revert changes in tests

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus committed:
Merge remote-tracking branch 'origin/add-username-to-events-updated' into add-username-to-events-updated

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge branch 'main' into add-username-to-events-updated

mpawlus committed:
test: adjust test

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Merge remote-tracking branch 'origin/add-username-to-events-updated' into add-username-to-events-updated

mpawlus committed:
fix: restore setting username on password reset

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

mpawlus committed:
Revert "test: adjust test"

This reverts commit 5c16d9905ec46e3b4231efa600ec54c9af42adcf.

Signed-off-by: Maciej Pawlus <m.t.pawlus@gmail.com>

stianst commented:
We could potentially add some way of showing additional information about the user, for example make the userid clickable to show a modal with details, but events are based on user-ids and not usernames, so simply replacing it is confusing and actually provides a worse UX since when searching for events it has to be based on the user-id.

mpawlus commented:
@stianst I understand the focus on user-id for search. However, the intent is schema consistency, as LOGIN events include the username while LOGOUT events do not. This missing data could affect external audit tools e.g. Splunk. I’ll move the discussion back to the issue https://github.com/keycloak/keycloak/issues/39357 to focus on data parity. Hope that helps clarify the intent!
